middleware/rate_limiter.go

```diff
@@ -0,0 +1,107 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// RateLimiterStore is the interface to be implemented by custom stores.
+type RateLimiterStore interface {
+	// Stores for the rate limiter have to implement the Allow method
+	Allow(identifier string) bool
+}
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// SourceFunc uses echo.Context to extract the identifier for a visitor
+		SourceFunc func(context echo.Context) string
+		// Store defines a store for the rate limiter
+		Store RateLimiterStore
+	}
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+	SourceFunc: func(ctx echo.Context) string {
+		return ctx.RealIP()
+	},
+}
+
+// RateLimiter returns a rate limiting middleware
+func RateLimiter(store RateLimiterStore) echo.MiddlewareFunc {
+	config := DefaultRateLimiterConfig
+	config.Store = store
+
+	return RateLimiterWithConfig(config)
+}
+
+// RateLimiterWithConfig returns a rate limiting middleware
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.SourceFunc == nil {
+		config.SourceFunc = DefaultRateLimiterConfig.SourceFunc
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier := config.SourceFunc(c)
+
+			allowed := config.Store.Allow(identifier)
+			if !allowed {
+				return c.JSON(http.StatusTooManyRequests, nil)
```

Comment:
It might me required to configure the HTTP code to return as other reverse proxies or handlers might rely on it.
Should we make that configurable?

Reply:
Makes sense! It shouldn't be fixed at a 429. Maybe an error handler can be set as a property of the RateLimiter itself. I'll do that and add it to the PR. Great suggestion!!